File: RandytheStudent V2.0/RandyTheStudent/RandyTheStudent/RandyMenu.m
//
//  RandyMenu.m
//  RandyTheStudent
//
//  Created by Enrique Galicia on 8/21/12.
//  Copyright (c) 2012 Enrique Galicia. All rights reserved.
//

#import "RandyMenu.h"
#import "Classes.h"


@implementation RandyMenu
@synthesize flipsidePopoverController2;

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}
-(NSString*)randy:(NSString*)key comment:(NSString*)comment{
    NSBundle *b =[NSBundle mainBundle];
    NSString *str =[b localizedStringForKey:key value:nil table:@"RandyLocal"];
    return str;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    sub=[[Subtitulados alloc]init];
    valores=[[NSMutableDictionary alloc]init];
    
    //Crear Campos de Base Principal
    NSArray *tablaarchivos=[NSArray arrayWithObjects:@"ARCHIVOS",@"NOMBRE TEXT",@"ARCHITEXT TEXT", nil];
    NSArray *concentrado=[NSArray arrayWithObjects:tablaarchivos, nil];
    
    //Crear Base Principal
    BasedeBases=[[DataBase alloc]initDB:@"Principal2.db" Tablas:concentrado];
    
    NSArray *BDPrincipal=[NSArray arrayWithObjects:@"Id",@"Nombre",@"Architext", nil];
    NSArray *BDPrincipalS=[NSArray arrayWithObjects:@"Nombre",@"Architext", nil];
    
    
    NSArray *Valores=[NSArray arrayWithObjects:@"Default",@"Default2.db", nil];
    [valores setObject:BDPrincipal forKey:@"BDPrincipal"];
    [valores setObject:BDPrincipalS forKey:@"BDPrincipalS"];
    
    
    if ([[[BasedeBases getalltablesfromDB:@"archivos" campos:BDPrincipal] objectForKey:@"Id"] count]==0) {
        [BasedeBases revisarBD:BDPrincipalS Valores:Valores Testigo:@"Nombre" Tabla:@"archivos" Campo:@"Nombre" Nombre:@"Default"];
    }
    
    prefs = [NSUserDefaults standardUserDefaults];
    NSString* strCat= [prefs objectForKey:@"Archivo"];
    if (!strCat) {
        [prefs setObject:@"Default2.db" forKey:@"Archivo"];
        [prefs synchronize];
    }
    strCat= [prefs objectForKey:@"Archivo"];
    [valores setObject:strCat forKey:@"NArchivo"];
     NSString *narchivo=[BasedeBases getselecteddatafromstatement:[NSString stringWithFormat:@"SELECT nombre FROM archivos WHERE architext=\"%@\"",[valores objectForKey:@"NArchivo"]]];
    
    
    BasesDatos=[[ComboBox alloc]init];
    BasesDatos.delegadocombo=self;
    [BasesDatos setActivo:1];
    BasesDatos.titulo=@"Archivo";
    BasesDatos.view.frame= IvBase.frame;
    [self.view addSubview:BasesDatos.view];
    [BasesDatos text:narchivo];

    [BasesDatos setComboData:[sub MAtitc1:[BasedeBases getalltablesfromDB:@"archivos" campos:BDPrincipal] c1:@"Nombre" ]];
    
    [BasesDatos text:[sub titc1:[BasedeBases getalldatafromstatement:[NSString stringWithFormat:@"SELECT * FROM archivos WHERE architext=\"%@\"",strCat] campos:BDPrincipal] c1:@"Nombre"]];
    [valores setObject:[sub titc1:[BasedeBases getalldatafromstatement:[NSString stringWithFormat:@"SELECT * FROM archivos WHERE architext=\"%@\"",strCat] campos:BDPrincipal] c1:@"Nombre"] forKey:@"NArchivo"];
    [valores setObject:[BasedeBases getselecteddatafromstatement:[NSString stringWithFormat:@"SELECT architext FROM archivos WHERE nombre=\"%@\"",[valores objectForKey:@"NArchivo"]]] forKey:@"Archivo"];
    
}
//Informacion del Seleccionado ComboBox
-(void)sendselection:(NSString*)sendselection titulo:(NSString*)titulo{
    //NSLog(@"seleccion %@_%@",sendselection,titulo);

    
    if ([titulo isEqualToString:@"Archivo"]) {
        [valores setObject:sendselection forKey:@"NArchivo"];
        NSString *archivo=[BasedeBases getselecteddatafromstatement:[NSString stringWithFormat:@"SELECT architext FROM archivos WHERE nombre=\"%@\"",[valores objectForKey:@"NArchivo"]]];
        [valores setObject:archivo forKey:@"Archivo"];
        [prefs setObject:archivo forKey:@"Archivo"];
        [prefs synchronize];
        NuevaClase.text=sendselection;
    }
}

-(void)ClassesDidFinish:(Classes*)controller{
    [self dismissViewControllerAnimated:YES completion:nil];
}




- (void)viewDidUnload
{
    [super viewDidUnload];

}

-(IBAction)enterclass:(id)sender{
    Classes *clase;
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        clase= [[Classes alloc] initWithNibName:@"MainViewController_iPhone" bundle:nil];
    } else {
        clase= [[Classes alloc] initWithNibName:@"MainViewController_iPad" bundle:nil];
    }

    clase.delegatem=self;
    clase.archivo1=[valores objectForKey:@"Archivo"];

    [self presentViewController:clase animated:YES completion:nil];
}
-(IBAction)newclass:(id)sender{
    if (![NuevaClase.text isEqualToString:@""]) {
        NSArray *informacion=[NSArray arrayWithObjects:NuevaClase.text,[NSString stringWithFormat:@"%@.db",NuevaClase.text],nil];
        [BasedeBases revisarBD:[valores objectForKey:@"BDPrincipalS"] Valores:informacion Testigo:@"nombre" Tabla:@"archivos" Campo:@"nombre" Nombre:NuevaClase.text];
        [BasesDatos setComboData:[sub MAtitc1:[BasedeBases getalltablesfromDB:@"archivos" campos:[valores objectForKey:@"BDPrincipal"]] c1:@"Nombre" ]];
    }
}
-(IBAction)renameclass:(id)sender{
    if (![NuevaClase.text isEqualToString:@""]) {
        NSString *idd=[BasedeBases getselecteddatafromstatement:[NSString stringWithFormat:@"SELECT id FROM archivos WHERE nombre=\"%@\"",BasesDatos.selectedText]];
        [BasedeBases update:[NSString stringWithFormat:@"UPDATE archivos SET nombre =\"%@\" WHERE id =\"%@\"",NuevaClase.text,idd]];
        [BasesDatos setComboData:[sub MAtitc1:[BasedeBases getalltablesfromDB:@"archivos" campos:[valores objectForKey:@"BDPrincipal"]] c1:@"Nombre" ]];
        
        [valores setObject:NuevaClase.text forKey:@"NArchivo"];
        NSString *archivo=[BasedeBases getselecteddatafromstatement:[NSString stringWithFormat:@"SELECT architext FROM archivos WHERE nombre=\"%@\"",[valores objectForKey:@"NArchivo"]]];
        [valores setObject:archivo forKey:@"Archivo"];
        [prefs setObject:archivo forKey:@"Archivo"];
        [prefs synchronize];
        NuevaClase.text=NuevaClase.text;
        [BasesDatos text:NuevaClase.text];
        
        
        
    }
    
    
    
}

-(void)CreditsDidFinish:(Credits*)controller{
    NSLog(@"Ejecute Cierre");
    [self.flipsidePopoverController2 dismissPopoverAnimated:YES];
}
-(IBAction)credits:(id)sender{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        Credits *credits = [[Credits alloc] initWithNibName:@"Credits" bundle:nil];
        credits.delegatec = self;
        [self presentViewController:credits animated:YES completion:nil];
    } else {
        if (!self.flipsidePopoverController2) {
            Credits *credits = [[Credits alloc] initWithNibName:@"Credits" bundle:nil];
            credits.delegatec = self;
            self.flipsidePopoverController2 = [[UIPopoverController alloc] initWithContentViewController:credits];
        }
        if ([self.flipsidePopoverController2 isPopoverVisible]) {
            [self.flipsidePopoverController2 dismissPopoverAnimated:YES];
        } else {
            [self.flipsidePopoverController2 presentPopoverFromRect:CGRectMake(0,0,320,480) inView:self.view permittedArrowDirections:UIPopoverArrowDirectionRight animated:YES];
        }
    }
}
-(void)MainDidFinish:(Classes*)controller{
    [self dismissViewControllerAnimated:NO completion:nil];
}


@end

File: RandytheStudent V2.0/RandyTheStudent/RandyTheStudent/RandyMenu.swift
//
//  RandyMenu.swift
//  RandyTheStudent
//
//  Swift port of the original RandyMenu.m (Objective-C) — the main menu
//  screen (pick/create/rename a class, open it, show credits).
//  `@objc(RandyMenu)` is required: both RandyMenu.xib (iPad) and
//  RandyMenu_Iphone.xib load via the "nib name matches class name"
//  convention, and AppDelegate.m (still Objective-C until it's converted
//  too) instantiates this class directly by name.
//

import UIKit

@objc(RandyMenu)
class RandyMenu: UIViewController, ComboDelegate, ClassesDelegate, CreditsDelegate {

    @IBOutlet private weak var lclases: UIButton!
    @IBOutlet private weak var nuevaClase: UITextField!
    @IBOutlet private weak var nueva: UIButton!
    @IBOutlet private weak var rename: UIButton!
    @IBOutlet private weak var creditsButton: UIButton!
    @IBOutlet private weak var ivBase: UIImageView!

    private var basedeBases: DataBase!
    private var basesDatos: ComboBox!
    private var valores = NSMutableDictionary()
    private var prefs: UserDefaults!
    private var sub: Subtitulados!

    private func randy(_ key: String, comment: String) -> String {
        Bundle.main.localizedString(forKey: key, value: nil, table: "RandyLocal")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        // Force Auto Layout to resolve ivBase's frame now, since it gets copied
        // into basesDatos.view.frame below before the deferred layout pass would run.
        view.layoutIfNeeded()
        sub = Subtitulados()
        valores = NSMutableDictionary()

        // Crear Campos de Base Principal
        let tablaarchivos = ["ARCHIVOS", "NOMBRE TEXT", "ARCHITEXT TEXT"]
        let concentrado = [tablaarchivos]

        // Crear Base Principal
        basedeBases = DataBase(file: "Principal2.db", tablas: concentrado as NSArray)

        let bdPrincipal: NSArray = ["Id", "Nombre", "Architext"]
        let bdPrincipalS: NSArray = ["Nombre", "Architext"]

        let valoresIniciales: NSArray = ["Default", "Default2.db"]
        valores["BDPrincipal"] = bdPrincipal
        valores["BDPrincipalS"] = bdPrincipalS

        let existingIds = (basedeBases.getalltablesfromDB("archivos", campos: bdPrincipal)["Id"] as? [Any]) ?? []
        if existingIds.isEmpty {
            basedeBases.revisarBD(bdPrincipalS, valores: valoresIniciales, testigo: "Nombre", tabla: "archivos", campo: "Nombre", nombre: "Default")
        }

        prefs = UserDefaults.standard
        var strCat = prefs.string(forKey: "Archivo")
        if strCat == nil {
            prefs.set("Default2.db", forKey: "Archivo")
        }
        strCat = prefs.string(forKey: "Archivo")
        valores["NArchivo"] = strCat
        let narchivo = basedeBases.getselecteddatafromstatement("SELECT nombre FROM archivos WHERE architext=\"\(valores["NArchivo"] ?? "")\"")

        basesDatos = ComboBox()
        basesDatos.delegadocombo = self
        basesDatos.activo = 1
        basesDatos.titulo = "Archivo"
        // Converted (not read directly) because ivBase now sits inside the
        // centered layout wrapper added to the XIB, so its .frame is in that
        // wrapper's coordinate space, not view's.
        basesDatos.view.frame = ivBase.superview?.convert(ivBase.frame, to: view) ?? ivBase.frame
        view.addSubview(basesDatos.view)
        basesDatos.text(narchivo)

        basesDatos.setComboData(sub.MAtitc1(basedeBases.getalltablesfromDB("archivos", campos: bdPrincipal), c1: "Nombre"))

        let datosArchivo = basedeBases.getalldatafromstatement("SELECT * FROM archivos WHERE architext=\"\(strCat ?? "")\"", campos: bdPrincipal)
        let nombreArchivo = sub.titc1(datosArchivo, c1: "Nombre")
        basesDatos.text(nombreArchivo)
        valores["NArchivo"] = nombreArchivo
        valores["Archivo"] = basedeBases.getselecteddatafromstatement("SELECT architext FROM archivos WHERE nombre=\"\(valores["NArchivo"] ?? "")\"")
    }

    // MARK: - Informacion del Seleccionado ComboBox

    func sendselection(_ sendselection: String, titulo: String) {
        guard titulo == "Archivo" else { return }
        valores["NArchivo"] = sendselection
        let archivo = basedeBases.getselecteddatafromstatement("SELECT architext FROM archivos WHERE nombre=\"\(valores["NArchivo"] ?? "")\"")
        valores["Archivo"] = archivo
        prefs.set(archivo, forKey: "Archivo")
        nuevaClase.text = sendselection
    }

    @objc(ClassesDidFinish:)
    func classesDidFinish(_ controller: Classes!) {
        dismiss(animated: true, completion: nil)
    }

    @IBAction private func enterclass(_ sender: Any) {
        let clase: Classes
        if UIDevice.current.userInterfaceIdiom == .phone {
            clase = Classes(nibName: "MainViewController_iPhone", bundle: nil)
        } else {
            clase = Classes(nibName: "MainViewController_iPad", bundle: nil)
        }

        clase.delegatem = self
        clase.archivo1 = valores["Archivo"] as? String

        present(clase, animated: true, completion: nil)
    }

    @IBAction private func newclass(_ sender: Any) {
        guard let text = nuevaClase.text, !text.isEmpty else { return }
        let informacion: NSArray = [text, "\(text).db"]
        basedeBases.revisarBD((valores["BDPrincipalS"] as? NSArray) ?? [], valores: informacion, testigo: "nombre", tabla: "archivos", campo: "nombre", nombre: text)
        basesDatos.setComboData(sub.MAtitc1(basedeBases.getalltablesfromDB("archivos", campos: (valores["BDPrincipal"] as? NSArray) ?? []), c1: "Nombre"))
    }

    @IBAction private func renameclass(_ sender: Any) {
        guard let text = nuevaClase.text, !text.isEmpty else { return }
        let idd = basedeBases.getselecteddatafromstatement("SELECT id FROM archivos WHERE nombre=\"\(basesDatos.selectedText)\"")
        basedeBases.update("UPDATE archivos SET nombre =\"\(text)\" WHERE id =\"\(idd)\"")
        basesDatos.setComboData(sub.MAtitc1(basedeBases.getalltablesfromDB("archivos", campos: (valores["BDPrincipal"] as? NSArray) ?? []), c1: "Nombre"))

        valores["NArchivo"] = text
        let archivo = basedeBases.getselecteddatafromstatement("SELECT architext FROM archivos WHERE nombre=\"\(valores["NArchivo"] ?? "")\"")
        valores["Archivo"] = archivo
        prefs.set(archivo, forKey: "Archivo")
        basesDatos.text(text)
    }

    func creditsDidFinish(_ controller: Credits) {
        dismiss(animated: true, completion: nil)
    }

    @IBAction private func credits(_ sender: Any) {
        if presentedViewController != nil {
            dismiss(animated: true, completion: nil)
            return
        }
        let credits = Credits(nibName: "Credits", bundle: nil)
        credits.delegatec = self
        if UIDevice.current.userInterfaceIdiom == .phone {
            present(credits, animated: true, completion: nil)
        } else {
            credits.modalPresentationStyle = .popover
            credits.popoverPresentationController?.sourceView = view
            credits.popoverPresentationController?.sourceRect = CGRect(x: 0, y: 0, width: 320, height: 480)
            credits.popoverPresentationController?.permittedArrowDirections = .right
            present(credits, animated: true, completion: nil)
        }
    }

    /// Not part of ClassesDelegate (only ClassesDidFinish: is) and never called
    /// from Classes — kept for parity with the original, which also never called it.
    func mainDidFinish(_ controller: Classes) {
        dismiss(animated: false, completion: nil)
    }
}
